systemFlake.nix

```
@@ -24,18 +25,34 @@
 }@args:
 
 let
+  # clean dirty api input / ensure for that all expected, but no extra attrs are present
+  validateHostAttrs = {
+      channelName ? null
+    , system ? null
+    , modules ? []
+    , extraArgs ? {}
+  } @ args: { inherit channelName system modules extraArgs; };
+
+  mergeHostAttrs = defaultAttrs: attrs: let
+    defaultTo = lhs: rhs: if rhs != null then rhs else lhs;
+  in
+  {
+    channelName = defaultTo (defaultTo "nixpkgs" defaultAttrs.channelName) attrs.channelName;
+    system = defaultTo (defaultTo defaultSystem defaultAttrs.system) attrs.system; # replace deaultSystem with x86_64-linux
+    modules = modules ++ defaultAttrs.modules ++ sharedModules;
+    extraArgs = sharedExtraArgs // defaultAttrs.extraArgs // extraArgs;
+  };
+
```

```suggestion
evalHostArgs = 
  { channelName ? "nixpkgs"
  , system ? defaultSystem
  , modules ? [ ]
  , extraArgs ? {}
  }:  
  { 
    inherit channelName system; 
    modules = sharedModules ++ defaultHsotAttrs.modules ++ modules;
    extraArgs = sharedExtraArgs // defaultHostAttrs.extraArgs // extraArgs;
  }
```
And `evalHostArgs` can still be called with `defaultHostAttrs // hostAttrs`, so // takes care of channelName and system and we can take care of modules/extraArgs manually, since those are a bit more nuanced.
To bring the discussion context here: I would prefer validation done in the outside of the function(arguments) and logic done on the inside. I'm not a fan of separating the two. @blaggacao wanted to separate the concerns to be clearer, because we was against doing logic like `channelName ? "nixpkgs"` in the same function that deals with merging. So his argument is validation and merging should not be done in the same function.

Overall goal here is to ensure contract validation for `defaultHostAttrs` and `hostAttrs`. I think both pieces of code accomplish that.

---

`defaultHostAttrs // hostAttrs` this not intended behaviour, actually.

I intentionally expect for:
```nix
{
  defaultHostAttrs.modules = [ a ];
  hostAttrs.modules = [ b ];
}
```
to become
```
{
  hostAttrs.module = [ a b ];
}
```

As far as I can understand the suggestion from merely reading it, we would arrive at:
```nix
{
  hostAttrs.module = [ b ];
}
```

That is not intended behavior of this PR. Am I missing something?

---

but then thats what these lines are for:
```nix
modules = sharedModules ++ defaultHsotAttrs.modules ++ modules;
extraArgs = sharedExtraArgs // defaultHostAttrs.extraArgs // extraArgs;
```
So the more intricate ones we merge manually. And importing modules twice isn't a concern, since module system deals with that.

---

So your right in that the `modules` that gets passed to `evalHostArgs` woould be `[ b ]` but then we do the merge again for modules, so it would return modules as `[ a b ]`.

---

Ah, now I think I spotted it. Is there a way to make this more obvious? (with a little less indirection and "double floors")

---

I'm not sure, I think this is the best way to do it. If you try and do it in the arguments, you end up with the problem that it gets overridden and not merged. I mean this is what you did in `mergeHostArgs` I'm just suggesting to do it in the same function as where defaults are set.

---

I mean is there just a way to _not_ tricking the reader into thinking that `rhs` will override `lhs` when doing `defaultHostAttrs // hostAttrs`? Like to accommodate a bit the average (non-expert) reader's attention span, that might just stop here and say: "aha, it's overridden then. Makes sense, it's called 'default', anyhow" (zoink!).

---

Ahh I think I finally now get your reasoning for `mergeHosts`. 
I think my suggestion is more accommodating to the reader that only looks at the top level attributes and `evalHostArgs`, its really clear whats going on, while your code requires a bit of hoops to jump through to see whats happening. 
Whereas for the reader at `defaultHostAttrs // hostAttrs` it might be confusing, but to be fair if you've read the top code I think it would be clear at this point. 
And I still think we should care more about the former reader than the latter, since most users and even contributors, likely won't go far enough into implementation to get confused by that one line.

---

I honestly did get confused. Now, I'm not sure what kind of read I am. :stuck_out_tongue_winking_eye: 

Let me see, what can be done.

---

@Pacman99

I think we can agree, that we want to avoid `defaultHostAttrs // hostAttrs` on the calling site, since it elicits a mislead interpretation of the code, a trap, we ourselves by the way have fallen into, previously.

Can you maybe adapt your suggestion accordingly to see if we can find a pareto-efficient solution to your concern?

---

So I addressed the awkward reading of `defaultTo`, which was plain bad English. I hope with this change, the code reads more easily, and the merge semantics are effortlessly evident to any class of reader. The reader is not expected to be forced to understand my "nix-kell" in-depth, but to take natural language which is construed with its help at face value.